### ISSUE TITLE
TypedDict with Union reports wrong types

### ISSUE THREAD

**Issue by bryanforbes** [python/mypy#5395]
Python version: 3.7
mypy version: 0.620
mypy flags: `--python-version 3.5 --disallow-untyped-defs --no-implicit-optional`

Given the following:

```python
PartialDict = TypedDict('PartialDict', {
    'ids': Iterable[str],
    'channel_id': str
})
FullDict = TypedDict('FullDict', {
    'ids': Iterable[str],
    'channel_id': str,
    'guild_id': str
})
def foo(data: Union[FullDict, PartialDict]) -> None:
    reveal_type(data['channel_id']) # revealed type is typing.Iterable*[builtins.str]
```

`reveal_type()` is reporting that `channel_id` is the same type as `ids`. The reason I'm using a `Union` is so I can write typings that will work in Python 3.5.3. For now, I've worked around this by using an overload:

```python
@overload
def foo(data: FullDict) -> None: ...
@overload
def foo(data: PartialDict) -> None: ...
def foo(data: Any) -> None:
    pass
```

But I don't get internal type checking within the function.

**Comment by gvanrossum** [comment]
Yeah, it looks like it falls back to `Dict` when inferring the type of `data['channel_id']`. To prove this, if you try `data['foo']` it gives the same result, and no error about an invalid key.

I suppose this could be fixed but given that you've got a work-around I'd say it's low priority. If you feel like it you could try to figure out where to add the special-casing to mypy and submit a PR though!

**Comment by ilevkivskyi** [comment]
This is now fixed on master.

**Comment by hevy-CC4477** [comment]
I have an issue that is related (with Mypy 720). I am trying to get an aws lambda handler to use two potential type of  TypedDict, one is a test event sent by aws another one is a put event sent by a s3 bucket.

```
S3ObjectEvent = TypedDict("S3ObjectEvent", {"Records": List[Any]})
S3TestEvent = TypedDict("S3TestEvent", {"Event": str})
LambdaEvent = Union[S3TestEvent, S3ObjectEvent]
```

I get the following error: `error: TypedDict "S3ObjectEvent" has no key 'Event'` with this function:
```
def is_test_event(event: LambdaEvent) -> bool:
    return event.get("Event", "") == "s3:TestEvent"
```

**Comment by hevy-CC4477** [comment]
forgot to include:
Python 3.7.4
mypy flags: `--strict`

**Comment by JukkaL** [comment]
Mypy is strictly speaking correct here, since TypedDicts use structural subtyping, so the key `'Event'` could be defined for `S3ObjectEvent` as well (but not visible through the type), and potentially with an incompatible type. There is an existing issue about this: #7151. Feel free to contribute that discussion.
